src/CSnakes.Runtime/CPython/Init.cs

```diff
@@ -65,7 +67,47 @@ internal void Initialize()
         if (IsInitialized)
             return;
 
-        InitializeEmbeddedPython();
+        /* Notes:
+         * 
+         * The CPython initialization and finalization
+         * methods should be called from the same thread.
+         * 
+         * Without doing so, CPython can hang on finalization.
+         * Especially if the code imports the threading module, or 
+         * uses asyncio.
+         * 
+         * Since we have no guarantee that the Dispose() of this 
+         * class will be called from the same managed CLR thread
+         * as the one which called Init(), we create a Task with 
+         * a cancellation token and use that as a mechanism to wait
+         * in the background for shutdown then call the finalization.
+         */
+
+        cts = new CancellationTokenSource();
+        bool initialized = false;
+        initializationTask = Task.Run(
+            () => {
+                InitializeEmbeddedPython();
+                initialized = true;
+                while (true)
+                {
+                    if (cts.IsCancellationRequested)
+                    {
+                        FinalizeEmbeddedPython();
+                        break;
+                    }
+                    else
+                    {
+                        Thread.Sleep(500);
+                    }
+                }
+            },
+            cancellationToken: cts.Token);
+
+        while (!initialized) // Wait for startup
+        {
+            continue;

```

Comment:
I'm sure there's a better pattern for doing this, also we should maybe add a 50ms sleep here